### PR TITLE
implement `/database` and `/database/query`

### PR DESCRIPTION
**Pull request by harryob**
hopefully! this probably needs work and a more comprehensive test suite - `/database/query/Reset()` isn't implemented because it seems a real ass

### CODE REVIEW

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    public override void Initialize(DreamProcArguments args) {
+        base.Initialize(args);
+
+        if (!args.GetArgument(0).TryGetValueAsString(out var command)) {
+            return;
+        }
+
+        SetupCommand(command, args.Values[1..]);
+
+    }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:23]:
```suggestion
    }
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    /// Sets up the SQLiteCommand, setting up parameters when provided.
+    /// Supports strings and floats from DMcode.
+    /// </summary>
+    /// <param name="command">The command text of the SQLite command, with placeholders denoted by '?'</param>
+    /// <param name="values">The values to be substituted into the command</param>
+    public void SetupCommand(string command, ReadOnlySpan<DreamValue> values) {
+        _command = new SqliteCommand(ParseCommandText(command));
+
+        for (var i = 0; i < values.Length; i++) {
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:40]:
```suggestion
        for (var i = 0; i < values.Length; i++) {
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            var arg = values[i];
+
+            var type = arg.Type;
+            switch (type) {
+                case DreamValue.DreamValueType.String:
+                    if (arg.TryGetValueAsString(out var stringValue)) {
+                        _command.Parameters.AddWithValue($"@{i}", stringValue);
+                    }
+                    break;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:49]:
```suggestion
                    }
                    
                    break;
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                case DreamValue.DreamValueType.Float:
+                    if (arg.TryGetValueAsFloat(out var floatValue)) {
+                        _command.Parameters.AddWithValue($"@{i}", floatValue);
+                    }
+                    break;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:54]:
```suggestion
                    }
                    
                    break;
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                case DreamValue.DreamValueType.DreamResource:
+                case DreamValue.DreamValueType.DreamObject:
+                case DreamValue.DreamValueType.DreamType:
+                case DreamValue.DreamValueType.DreamProc:
+                case DreamValue.DreamValueType.Appearance:
+                default:
+                    // TODO: support saving BLOBS for icons, if we really want to
+                    break;
+            }
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:65]:
```suggestion
            }
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        }
+
+        var names = new List<DreamValue>();
+        for (var i = 0; i < _reader.FieldCount; i++) {
+            names.Add(new DreamValue(_reader.GetName(i)));
+        }
+
+        return names;
+
+    }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:85]:
```suggestion
    }
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            var name = _reader.GetName(id);
+            return new DreamValue(name);
+        } catch (IndexOutOfRangeException exception) {
+            _errorCode = 1;
+            _errorMessage = exception.Message;
+        }
+
+        return DreamValue.Null;
+
+    }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:107]:
```suggestion
    }
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        _command.Connection = connection;
+
+        try {
+            _reader = _command.ExecuteReader();
+        } catch (SqliteException exception) {
+            _errorCode = exception.SqliteErrorCode;
+            _errorMessage = exception.Message;
+            database.SetError(exception.SqliteErrorCode, exception.Message);
+        }
+        ClearCommand();
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:147]:
```suggestion
        }
        
        ClearCommand();
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        };
+    }
+
+    /// <summary>
+    /// Builds a new string, converting '?' characters to expressions we can bind to later
+    /// </summary>
+    /// <param name="text">The raw command text</param>
+    /// <returns>A <see cref="string"/> with the characters converted</returns>
+    private static string ParseCommandText(string text) {
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:227]:
```suggestion
    private static string ParseCommandText(string text) {
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        var newString = new StringBuilder();
+
+        var paramsId = 0;
+        var inQuotes = false;
+        foreach (var character in text) {
+            switch (character)
+            {
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:234]:
```suggestion
            switch (character) {
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs [hunk 11]
```diff
@@ -0,0 +1,145 @@
+using OpenDreamRuntime.Objects;
+using OpenDreamRuntime.Objects.Types;
+using DreamValueTypeFlag = OpenDreamRuntime.DreamValue.DreamValueTypeFlag;
+
+namespace OpenDreamRuntime.Procs.Native;
+
+internal static class DreamProcNativeDatabaseQuery {
+
+    [DreamProc("Add")]
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs:9]:
```suggestion
internal static class DreamProcNativeDatabaseQuery {
    [DreamProc("Add")]
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+        var list = bundle.ObjectTree.CreateList();
+
+        foreach (var value in query.GetAllColumns()) {
+            list.AddValue(value);
+        }
+
+        return new DreamValue(list);
+
+    }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs:61]:
```suggestion
        return new DreamValue(list);
    }
```

File: OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        var list = bundle.ObjectTree.CreateList();
+        foreach (var entry in data) {
+            list.SetValue(new DreamValue(entry.Key), entry.Value);
+        }
+
+        return new DreamValue(list);
+
+    }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeDatabaseQuery.cs:143]:
```suggestion
        return new DreamValue(list);
    }
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    /// <summary>
+    /// Builds a new string, converting '?' characters to expressions we can bind to later
+    /// </summary>
+    /// <param name="text">The raw command text</param>
+    /// <returns>A <see cref="string"/> with the characters converted</returns>
+    private static string ParseCommandText(string text) {
+
+        var newString = new StringBuilder();
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectDatabaseQuery.cs:228]:
```suggestion
    private static string ParseCommandText(string text) {
        var newString = new StringBuilder();
```

File: OpenDreamRuntime/Objects/Types/DreamObjectDatabase.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    /// Establish the connection to our SQLite database
+    /// </summary>
+    /// <param name="filename">The path to the SQLite file</param>
+    public void Open(string filename) {
+
+        if (_connection?.State == ConnectionState.Open) {
+            Close();
+        }
+
+        _connection = new SqliteConnection($"Data Source={filename};Mode=ReadWriteCreate");
```
[the remark this answers is not in the document]
Comment (harryob) [reply OpenDreamRuntime/Objects/Types/DreamObjectDatabase.cs:42]: that's a silly one - addressed!

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**